File: src/widget/layout.c
#ifdef _PROTOTYPE_

#include "../land.h"
#include "base.h"

#endif /* _PROTOTYPE_ */

#include "widget/layout.h"

void land_widget_layout_inhibit(LandWidget *self)
{
    self->no_layout = 1;
}

void land_widget_layout_enable(LandWidget *self)
{
    self->no_layout = 0;
}

void land_widget_layout_set_grid(LandWidget *self, int columns, int rows)
{
    self->box.rows = rows;
    self->box.cols = columns;
    if (!self->no_layout)
        gul_layout_changed(&self->box);
}

void land_widget_layout_set_grid_position(LandWidget *self, int column, int row)
{
    self->box.col = column;
    self->box.row = row;
    if (self->parent && !self->parent->no_layout)
        gul_layout_changed(&self->parent->box);
}

void land_widget_layout_set_grid_extra(LandWidget *self, int columns, int rows)
{
    self->box.extra_cols = columns;
    self->box.extra_rows = rows;
    if (self->parent && !self->parent->no_layout)
        gul_layout_changed(&self->parent->box);
}

void land_widget_layout_set_minimum_size(LandWidget *self, int w, int h)
{
    self->box.min_width = w;
    self->box.min_height = h;
    if (self->box.current_min_width < w)
        self->box.current_min_width = w;
    if (self->box.current_min_height < 0)
        self->box.current_min_height = h;
    if (self->box.w < w)
        self->box.w = w;
    if (self->box.h < h)
        self->box.h = h;
}

void land_widget_layout_set_shrinking(LandWidget *self, int x, int y)
{
    if (x)
        self->box.flags |= GUL_SHRINK_X;
    if (y)
        self->box.flags |= GUL_SHRINK_Y;
}

void land_widget_layout_set_expanding(LandWidget *self, int x, int y)
{
    if (x)
        self->box.flags &= ~GUL_SHRINK_X;
    if (y)
        self->box.flags &= ~GUL_SHRINK_Y;
}

void land_widget_layout_set_border(LandWidget *self, int l, int t, int r, int b, int hgap, int vgap)
{
    self->box.il = l;
    self->box.it = t;
    self->box.ir = r;
    self->box.ib = b;
    self->box.hgap = hgap;
    self->box.vgap = vgap;
}

void land_widget_layout_add(LandWidget *parent, LandWidget *child)
{
    gul_remove_child(&parent->box, &child->box, !parent->no_layout);
    gul_attach_child(&parent->box, &child->box, !parent->no_layout);
}

int land_widget_layout(LandWidget *self)
{
    if (!self->no_layout)
        gul_layout_changed(&self->box);
    return gul_box_fit_children(&self->box, 0, 0);
}

int land_widget_layout_adjust(LandWidget *self, int x, int y)
{
    if (!self->no_layout)
        gul_layout_changed(&self->box);
    return gul_box_fit_children(&self->box, x, y);
}

File: src/widget/layout.py
import land/land, base

# Returns 1 if layout is now inhibited, 0 if it already was.
def land_widget_layout_freeze(LandWidget *self) -> int:
    int nl = self.no_layout
    self.no_layout = 1
    return not nl

# Returns 1 if layout is now enabled, 0 if it already was.
def land_widget_layout_unfreeze(LandWidget *self) -> int:
    int nl = self.no_layout
    self.no_layout = 0
    return nl

def land_widget_layout_set_grid(LandWidget *self, int columns, rows):
    self.box.rows = rows
    self.box.cols = columns
    land_widget_layout(self)

def land_widget_layout_disable(LandWidget *self):
    self.box.flags |= GUL_NO_LAYOUT

def land_widget_layout_enable(LandWidget *self):
    self.box.flags &= ~GUL_NO_LAYOUT

def land_widget_layout_set_grid_position(LandWidget *self, int column, row):
    self.box.col = column
    self.box.row = row
    if self.parent:
        land_widget_layout(self.parent)

def land_widget_layout_set_grid_extra(LandWidget *self, int columns, rows):
    self.box.extra_cols = columns
    self.box.extra_rows = rows
    if self.parent:
        land_widget_layout(self.parent)

def land_widget_layout_set_minimum_size(LandWidget *self, int w, h):
    self.box.min_width = w
    self.box.min_height = h

def land_widget_layout_set_maximum_size(LandWidget *self, int w, h):
    self.box.max_width = w
    self.box.max_height = h

def land_widget_layout_set_shrinking(LandWidget *self, int x, int y):
    if x:
        self.box.flags |= GUL_SHRINK_X
    if y:
        self.box.flags |= GUL_SHRINK_Y
        
    if self.parent and not self->parent->no_layout:
        land_widget_layout(self)

def land_widget_layout_set_expanding(LandWidget *self, int x, int y):
    if x:
        self.box.flags &= ~GUL_SHRINK_X
    if y:
        self.box.flags &= ~GUL_SHRINK_Y

    if self.parent:
        land_widget_layout(self)

def land_widget_layout(LandWidget *self):
    if not self.no_layout:
        land_internal_gul_layout_updated(self)

def land_widget_layout_initialize(LandWidget *self, int x, y, w, h):
    land_internal_land_gul_box_initialize(&self.box)
    self.box.x = x
    self.box.y = y
    self.box.w = w
    self.box.h = h
